### DISCUSSION TITLE
Dependent resources Deployment reconciling repeatedly

### DISCUSSION THREAD

**Discussion by cldfzn** [operator-framework/java-operator-sdk#2416] [Q&A]
I'm using the SDK to create a custom resource which creates an ingress, service, and deployment. Similar to the webpage example. However what happens when I run the operator, it creates the resources just fine. It then does not match the desired deployment to the actual deployment and I get thousands of generations of the deployment. When running through the code, it seems like the matcher only prunes managed fields from the actual deployment. When passing a desired state to the SSA matcher, the provided DeploymentBuilder ends up with a few extra empty fields in the converted map and never matches. Am I doing something wrong with this implementation or do I need to somehow prune out all of the extra fields so it matches in the SSA matcher? Or write my own matcher?

**Comment by csviri** [comment]
Hi @cldfzn , could you pls create a simple reproducer so we can take a look?

It is very hard to figure out without code, and we hade some cases when the implementation of the desired was wrong (like based on actual state). So If you prepare a sample minimalistic project on Github that reproduces it, then we can check it very quickly.

**Comment by cldfzn** [comment]
[Archive.zip](https://github.com/user-attachments/files/15516843/Archive.zip)

provided a zip file of an example project, after creating a crd instance it goes through generations repeatedly.

**Reply by csviri** [reply]
Thank you will take a look, one another approach just for now you can try, is to set:

`previousAnnotationForDependentResourcesEventFiltering` to false, in configuration service:

https://github.com/operator-framework/java-operator-sdk/blob/3c58e03afbbd5f08fbbafc8447a48f725afc3483/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java#L385-L385

**Reply by csviri** [reply]
@cldfzn can you send me a sample resource pls. thx

**Reply by cldfzn** [reply]
This will fail to start, but the deployment generations should kick off:

```
apiVersion: example.com/v1alpha1
kind: Frontend
metadata:
  name: hello
spec:
  image:
    registry: docker.io
  project: crccheck
  name: hello-world
  version: "v1.0.0"
  environment: test
```

**Reply by csviri** [reply]
Hi @cldfzn I tried your example, but was not able to reproduce this, it does not happens to, me on the attached file.

**Reply by cldfzn** [reply]
https://github.com/operator-framework/java-operator-sdk/assets/1036914/a18cf10e-c599-4dcb-a8d0-5d07c95709be

With the build I sent you, this happens. not sure what it would be. You can see the generations increasing pretty rapidly.

**Reply by csviri** [reply]
will try to take an another look next week.

Did you try this:
`previousAnnotationForDependentResourcesEventFiltering` to set into false in config service?

**Reply by cldfzn** [reply]
If I override the match function of DeploymentDependentResource with a true result the generations stop.

The values I see in the SSA matcher are:

prunedActual:
```
[{
    annotations={javaoperatorsdk.io/previous=479da2c0-abbc-4a79-87c0-d81f32718cff,9099599},
    labels={
        app.kubernetes.io/name=hello-world,
        app.kubernetes.io/instance=crccheck,
        app.kubernetes.io/version=v1.0.0,
        example.com/managed=true
    },
    ownerReferences=[{name=crccheck, uid=0f9fa9f2-a725-4158-889e-27e7e6261c15, apiVersion=example.com/v1alpha1, kind=PlatformFrontend}]
},
{
    template={
        metadata={
            labels={
                app.kubernetes.io/name=hello-world,
                app.kubernetes.io/instance=crccheck,
                app.kubernetes.io/version=v1.0.0,
                example.com/managed=true
            }
        },
        spec={
            tolerations=[],
            overhead={},
            readinessGates=[],
            serviceAccountName=default,
            containers=[{
                args=[],
                image=docker.io/crcheck/hello-world:v1.0.0,
                imagePullPolicy=IfNotPresent,
                name=app,
                readinessProbe={failureThreshold=6, periodSeconds=5, timeoutSeconds=1, successThreshold=1, initialDelaySeconds=1, httpGet={path=/apps/hello-world/readyz, scheme=HTTP, port=http, httpHeaders=[]}},
                resources={requests={}, limits={}},
                resizePolicy=[],
                securityContext={runAsUser=1000, capabilities={add=[], drop=[ALL]}, seccompProfile={type=RuntimeDefault}, allowPrivilegeEscalation=false, runAsGroup=1000, runAsNonRoot=true, readOnlyRootFilesystem=false},
                env=[{name=BASE_PATH, value=/apps/hello-world}],
                ports=[{protocol=TCP, name=http, containerPort=8080}],
                envFrom=[],
                command=[]
            }],
            securityContext={fsGroup=1000, supplementalGroups=[], sysctls=[]},
            nodeSelector={}
        }
    },
    replicas=1,
    selector={matchExpressions=[], matchLabels={app.kubernetes.io/instance=crccheck, app.kubernetes.io/name=hello-world}},
    strategy={type=Recreate}
}]
```

desiredMap:
```
[{
    annotations={javaoperatorsdk.io/previous=479da2c0-abbc-4a79-87c0-d81f32718cff,9099599},
    labels={
        app.kubernetes.io/instance=crccheck,
        app.kubernetes.io/name=hello-world,
        app.kubernetes.io/version=v1.0.0,
        example.com/managed=true,
    },
    ownerReferences=[{apiVersion=example.com/v1alpha1, kind=PlatformFrontend, name=crccheck, uid=0f9fa9f2-a725-4158-889e-27e7e6261c15}]
},
{
    replicas=1,
    selector={matchExpressions=[], matchLabels={app.kubernetes.io/instance=crccheck, app.kubernetes.io/name=hello-world}},
    strategy={type=Recreate},
    template={
        metadata={
            labels={
                app.kubernetes.io/instance=crccheck,
                app.kubernetes.io/name=hello-world,
                app.kubernetes.io/version=v1.0.0,
                example.com/managed=true,
            }
        },
        spec={
            containers=[{
                args=[],
                command=[],
                env=[{name=BASE_PATH, value=/apps/hello-world}],
                envFrom=[],
                image=docker.io/crcheck/hello-world:v1.0.0,
                imagePullPolicy=IfNotPresent,
                name=app,
                ports=[{containerPort=8080, name=http, protocol=TCP}],
                readinessProbe={failureThreshold=6, httpGet={httpHeaders=[], path=/apps/hello-world/readyz, port=http, scheme=HTTP}, initialDelaySeconds=1, periodSeconds=5, successThreshold=1, timeoutSeconds=1},
                resizePolicy=[],
                resources={claims=[], limits={}, requests={}},
                securityContext={allowPrivilegeEscalation=false, capabilities={add=[], drop=[ALL]}, readOnlyRootFilesystem=false, runAsGroup=1000, runAsNonRoot=true, runAsUser=1000, seccompProfile={type=RuntimeDefault}},
                volumeDevices=[],
                volumeMounts=[]
            }],
            ephemeralContainers=[],
            hostAliases=[],
            imagePullSecrets=[],
            initContainers=[],
            nodeSelector={},
            overhead={},
            readinessGates=[],
            resourceClaims=[],
            schedulingGates=[],
            securityContext={fsGroup=1000, supplementalGroups=[], sysctls=[]},
            serviceAccountName=default,
            tolerations=[],
            topologySpreadConstraints=[],
            volumes=[]
        }
    }
}]
```

These values can never match since the desired map has more fields in it, as far as I understand.

**Reply by cldfzn** [reply]
If I take the current SSABasedGenericKubernetesResourceMatcher implementation and modify it to do the same pruning to the desired map, I no longer have this issue:

```
val desiredActual = HashMap<String, Any?>(desiredMap.size)
keepOnlyManagedFields(desiredActual, desiredMap, managedFieldsEntry.fieldsV1.additionalProperties, objectMapper)
        
return prunedActual == desiredActual
```

**Reply by csviri** [reply]
So I tried it again, if I donwload the zip file run it from intellij and appyt the resource, the generation does not increase me. If you can describe the steps (if there are others) I can take a look.

**Reply by csviri** [reply]
However noticed something:

```
  val existing = context.getSecondaryResource(Deployment::class.java).orElse(null)
        return if (existing == null) {
            DeploymentBuilder().apply(primary).build()
        } else {
            DeploymentBuilder().apply(primary).build()
        }
```

This does not make sense to me. But I assume that you use the `existing` in you code. Note that this is very dangerous, the desired state should always use a fresh copy (not modifying the existing object), never have managed field and just express the desired state from you side.

**Reply by csviri** [reply]
> If I take the current SSABasedGenericKubernetesResourceMatcher implementation and modify it to do the same pruning to the desired map, I no longer have this issue:
> 
> ```
> val desiredActual = HashMap<String, Any?>(desiredMap.size)
> keepOnlyManagedFields(desiredActual, desiredMap, managedFieldsEntry.fieldsV1.additionalProperties, objectMapper)
>         
> return prunedActual == desiredActual
> ```

Since this works, I assume you have sometimes desired based on existing resource, try with a fresh copy if that works.

**Reply by cldfzn** [reply]
> However noticed something:
> 
> ```
>   val existing = context.getSecondaryResource(Deployment::class.java).orElse(null)
>         return if (existing == null) {
>             DeploymentBuilder().apply(primary).build()
>         } else {
>             DeploymentBuilder().apply(primary).build()
>         }
> ```
> 
> This does not make sense to me. But I assume that you use the `existing` in you code. Note that this is very dangerous, the desired state should always use a fresh copy (not modifying the existing object), never have managed field and just express the desired state from you side.

That was actually code to support some debugging. What I actually use is a clean desired state using the builder:

```
override fun desired(primary: Frontend, context: Context<Frontend>): Deployment =
        DeploymentBuilder()
            .withMetadata(metadata(primary))
            .withNewSpec()
            .withReplicas(1)
// etc...
```

**Reply by cldfzn** [reply]
The only process I have is building a native-image from the code using buildpacks and running it in EKS and rancher-desktop locally.

**Reply by csviri** [reply]
@metacosm can you take a look on this pls, if you can reproduce?

**Reply by cldfzn** [reply]
@csviri out of curiosity, how were you running the operator? did you use a native-image build or gradle or something else?

**Reply by csviri** [reply]
Was just running it from intellij.

**Reply by cldfzn** [reply]
Alright, well, that's the root cause of the issue. When I create a native-image from the operator code, this issue occurs. When I create a JVM application, things work as expected.

**Reply by csviri** [reply]
@cldfzn that makes sense. Could you pls convert this discussion to an issue either in JOSDK or QOSDK, also if you can take a look deeper what is happening would be good help, otherwise we will take a look and fix it eventually

**Reply by csviri** [reply]
cc @metacosm
